### PR TITLE
Translation memory

### PR DESCRIPTION
**Pull request by camertron**
By grabbing all unique meta keys and uploading them to smartling, we can make our own translation “memory” that contains all potential translations. The memory is essentially just a large in-memory hash that makes translation lookup really fast, meaning we can do pulls much, much faster than before.

### CODE REVIEW

File: lib/rosette/integrations/smartling-integration/smartling_completer.rb [hunk 1]
```diff
@@ -62,6 +62,30 @@ def delete_file(file)
...
+          rosette_config.datastore.add_or_update_commit_log(
+            repo_config.name, commit_id, nil, status
+          )
+
+          update_locale_logs(locales_map)
+        end
+
+        def derive_commit_id_from(locales_map)
+          repo_config.locales.each do |locale|
+            if file = locales_map[locale.code]
```
Comment (11mdlow) [inline lib/rosette/integrations/smartling-integration/smartling_completer.rb:84]: is this supposed to be assignment or equality?

File: lib/rosette/integrations/smartling-integration/smartling_completer.rb [hunk 2]
```diff
@@ -150,7 +174,12 @@ def each_file_for(locale, &block)
             list = get_file_list(locale_code, counter)
 
             while list.size > 0
-              list.each(&block)
+              list.each do |file|
+                if file.repo_name == repo_config.name
+                  yield file
```
Comment (11mdlow) [inline lib/rosette/integrations/smartling-integration/smartling_completer.rb]: this is pretty short, so let's put it on one line

File: lib/rosette/integrations/smartling-integration/smartling_puller.rb [hunk 3]
```diff
@@ -56,15 +56,16 @@ def pull
         private
 
         def pull_synchronously
-          status = status = Rosette::DataStores::PhraseStatus::PENDING
```
Comment (11mdlow) [inline lib/rosette/integrations/smartling-integration/smartling_puller.rb:59]: whoops

File: lib/rosette/integrations/smartling-integration/translation_memory.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        def initialize(translation_hash)
+          @translation_hash = translation_hash
+        end
+
+        def translation_for(locale, meta_key)
+          all_for_locale = @translation_hash.fetch(locale.code, DEFAULT_HASH)
+
+          if translation = all_for_locale[meta_key]
+            translation.key
+          end
```
Comment (11mdlow) [inline lib/rosette/integrations/smartling-integration/translation_memory.rb:19]: what will this return if there is no translation? maybe we should do: `all_for_locale[meta_key].try(:key)`?

File: lib/rosette/integrations/smartling-integration/smartling_completer.rb (hunk 1 above)
Comment (camertron) [inline lib/rosette/integrations/smartling-integration/smartling_completer.rb:84]: It's supposed to be assignment, the next line returns `file.commit_id`. If the result of the assignment is `nil`, the method looks at the other files in `locales_map`.

File: lib/rosette/integrations/smartling-integration/translation_memory.rb (hunk 4 above)
Comment (camertron) [inline lib/rosette/integrations/smartling-integration/translation_memory.rb:19]: If no translation is found, `#translation_for` returns a `nil`, which is what we want I think.

**Comment by 11mdlow** [comment]
:shipit:

**Pull request merged**